File: garudadidada.dart
/// Tiny zero-dependency helpers to humanize values:
/// byte sizes, durations, English ordinals, and pluralization.
library garudadidada;

const List<String> _byteUnits = [
  'KB',
  'MB',
  'GB',
  'TB',
  'PB',
  'EB',
  'ZB',
  'YB'
];

String _trim(double n) {
  var s = n.toStringAsFixed(1);
  if (s.endsWith('.0')) {
    s = s.substring(0, s.length - 2);
  }
  return s;
}

/// Format a byte count as a human-readable string using binary (1024) units.
///
/// ```dart
/// formatBytes(1536); // '1.5 KB'
/// ```
String formatBytes(int bytes) {
  var n = bytes.toDouble();
  if (n < 1024) {
    return '$bytes B';
  }
  for (var i = 0; i < _byteUnits.length; i++) {
    n /= 1024.0;
    if (n < 1024 || i == _byteUnits.length - 1) {
      return '${_trim(n)} ${_byteUnits[i]}';
    }
  }
  return '${_trim(n)} YB';
}

/// Format a duration in seconds as a human-readable string.
///
/// ```dart
/// formatDuration(3661); // '1h 1m 1s'
/// ```
String formatDuration(int seconds) {
  if (seconds == 0) {
    return '0s';
  }
  final parts = <String>[];
  final days = seconds ~/ 86400;
  seconds %= 86400;
  final hours = seconds ~/ 3600;
  seconds %= 3600;
  final minutes = seconds ~/ 60;
  seconds %= 60;
  if (days > 0) parts.add('${days}d');
  if (hours > 0) parts.add('${hours}h');
  if (minutes > 0) parts.add('${minutes}m');
  if (seconds > 0) parts.add('${seconds}s');
  return parts.join(' ');
}

/// Return [n] with its English ordinal suffix.
///
/// ```dart
/// ordinal(21); // '21st'
/// ```
String ordinal(int n) {
  var suffix = 'th';
  final m = n % 100;
  if (m < 11 || m > 13) {
    switch (n % 10) {
      case 1:
        suffix = 'st';
        break;
      case 2:
        suffix = 'nd';
        break;
      case 3:
        suffix = 'rd';
        break;
    }
  }
  return '$n$suffix';
}

/// Return `"<count> <singular|plural>"`, choosing the form based on [count].
///
/// ```dart
/// pluralize(3, 'file'); // '3 files'
/// ```
String pluralize(int count, String singular, [String? plural]) {
  plural ??= '${singular}s';
  final word = count == 1 ? singular : plural;
  return '$count $word';
}

File: garudadidada.rb
# frozen_string_literal: true

require_relative "garudadidada/version"

# Tiny zero-dependency helpers to humanize values:
# byte sizes, durations, English ordinals, and pluralization.
module Garudadidada
  BYTE_UNITS = %w[KB MB GB TB PB EB ZB YB].freeze

  module_function

  # Format a byte count as a human-readable string using binary (1024) units.
  #   Garudadidada.format_bytes(1536) # => "1.5 KB"
  def format_bytes(bytes)
    n = bytes.to_f
    return "#{bytes.to_i} B" if n < 1024

    last = BYTE_UNITS.length - 1
    BYTE_UNITS.each_with_index do |unit, i|
      n /= 1024.0
      return "#{trim(n)} #{unit}" if n < 1024 || i == last
    end
  end

  # Format a duration in seconds as a human-readable string.
  #   Garudadidada.format_duration(3661) # => "1h 1m 1s"
  def format_duration(seconds)
    s = seconds.to_i
    return "0s" if s.zero?

    parts = []
    days, s = s.divmod(86_400)
    hours, s = s.divmod(3_600)
    minutes, s = s.divmod(60)
    parts << "#{days}d" if days.positive?
    parts << "#{hours}h" if hours.positive?
    parts << "#{minutes}m" if minutes.positive?
    parts << "#{s}s" if s.positive?
    parts.join(" ")
  end

  # Return n with its English ordinal suffix.
  #   Garudadidada.ordinal(21) # => "21st"
  def ordinal(num)
    num = num.to_i
    m = num % 100
    suffix = if m.between?(11, 13)
               "th"
             else
               { 1 => "st", 2 => "nd", 3 => "rd" }.fetch(num % 10, "th")
             end
    "#{num}#{suffix}"
  end

  # Return "<count> <singular|plural>" choosing the form based on count.
  #   Garudadidada.pluralize(3, "file") # => "3 files"
  def pluralize(count, singular, plural = nil)
    plural ||= "#{singular}s"
    word = count == 1 ? singular : plural
    "#{count} #{word}"
  end

  def trim(num)
    format("%.1f", num).sub(/\.0$/, "")
  end
  private_class_method :trim
end
